src/django_components/context.py

```diff
@@ -17,10 +16,50 @@
 
 
 _FILLED_SLOTS_CONTENT_CONTEXT_KEY = "_DJANGO_COMPONENTS_FILLED_SLOTS"
-_OUTER_CONTEXT_CONTEXT_KEY = "_DJANGO_COMPONENTS_OUTER_CONTEXT"
+_OUTER_ROOT_CTX_CONTEXT_KEY = "_DJANGO_COMPONENTS_OUTER_ROOT_CTX"
 _SLOT_COMPONENT_ASSOC_KEY = "_DJANGO_COMPONENTS_SLOT_COMP_ASSOC"
 
 
+def prepare_context(context: Context, outer_context: Optional[Context]) -> None:
+    """Initialize the internal context state."""
+    # This is supposed to run ALWAYS at Component.render
+    if outer_context is not None:
+        set_outer_root_context(context, outer_context)
+
+    # Initialize a dict that will hold all slot -> component associations
+    # within this rendering run.
+    # This is shared across the whole render chain, thus we set it only once.
+    if _SLOT_COMPONENT_ASSOC_KEY not in context:
+        context[_SLOT_COMPONENT_ASSOC_KEY] = {}
+
+    # If we're inside a forloop, we need to make a disposable copy of slot -> comp
+    # mapping, which can be modified in the loop. We do so by copying it onto the latest
+    # context layer.
+    #
+    # This is necessary, because otherwise if we have a nested loop with a same
+    # component used recursively, the inner slot -> comp mapping would leak into the outer.
+    #
+    # NOTE: If you ever need to debug this, insert a print/debug statement into
+    # `django.template.defaulttags.ForNode.render` to inspect the context object
+    # inside the for loop.
+    if "forloop" in context:
+        context.dicts[-1][_SLOT_COMPONENT_ASSOC_KEY] = context[_SLOT_COMPONENT_ASSOC_KEY].copy()
```

Comment:
One of the issues was with a for loop. Don't know exactly why, but when there was nested loops, then, after the inner loop finished, some state 
surrounding `_SLOT_COMPONENT_ASSOC_KEY` persisted after an inner loop finished.

So what was basically happening in the bug was that if we had nested loops with slots, then after the inner loops finished, then the next iteration of the outer loop rendered a slot of the INNER loop, instead of a slot of an OUTER loop.